src/main/java/com/gluonhq/substrate/util/windows/MSIBundler.java

```diff
@@ -0,0 +1,343 @@
+ */
+package com.gluonhq.substrate.util.windows;
+
+import com.gluonhq.substrate.Constants;
+import com.gluonhq.substrate.model.InternalProjectConfiguration;
+import com.gluonhq.substrate.model.ProcessPaths;
+import com.gluonhq.substrate.model.ReleaseConfiguration;
+import com.gluonhq.substrate.util.FileOps;
+import com.gluonhq.substrate.util.Logger;
+import com.gluonhq.substrate.util.ProcessRunner;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Writer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.gluonhq.substrate.model.ReleaseConfiguration.DEFAULT_APP_VERSION;
+
+public class MSIBundler {
+
+    private final ProcessPaths paths;
+    private final InternalProjectConfiguration projectConfiguration;
+    private final String sourceOS;
+    private final Path rootPath;
+
+    public MSIBundler(ProcessPaths paths, InternalProjectConfiguration projectConfiguration) {
+        this.paths = paths;
+        this.projectConfiguration = projectConfiguration;
+        this.sourceOS = projectConfiguration.getTargetTriplet().getOs();
+        this.rootPath = paths.getSourcePath().resolve(sourceOS);
+    }
+
+    public boolean createPackage(boolean sign) throws IOException, InterruptedException {
+
+        if (WixTool.CANDLE.getPath() == null || WixTool.LIGHT.getPath() == null) {
+            Logger.logSevere("Please make sure to install Wix Toolset (https://wixtoolset.org/) before proceeding.");
+            return false;
+        }
+
+        final String appName = projectConfiguration.getAppName();
+        final String version = Optional.ofNullable(projectConfiguration.getReleaseConfiguration().getVersionName()).orElse(DEFAULT_APP_VERSION);
+        Path localAppPath = paths.getAppPath().resolve(appName + ".exe");
+        if (!Files.exists(localAppPath)) {
+            throw new IOException("Error: " + appName + ".exe not found");
+        }
+        Logger.logInfo("Building exe for " + localAppPath);
+
+        /**
+         * Create directories and copy resources
+         */
+        Path tmpMSI = paths.getTmpPath().resolve("tmpMSI");
+        if (Files.exists(tmpMSI)) {
+            FileOps.deleteDirectory(tmpMSI);
+        }
+        Files.createDirectories(tmpMSI);
+
+        Path windowsGenSrcPath = paths.getGenPath().resolve(sourceOS);
+        Path windowsAssetPath = windowsGenSrcPath.resolve(Constants.WIN_ASSETS_FOLDER);
```

Comment:
This can be moved below, since it is only used if (iconPath == null || !Files.exists(iconPath)):

```
if (iconPath == null || !Files.exists(iconPath)) {
    Path windowsGenSrcPath = paths.getGenPath().resolve(sourceOS);
    Path windowsAssetPath = windowsGenSrcPath.resolve(Constants.WIN_ASSETS_FOLDER);
    Files.createDirectories(windowsAssetPath);
...
}

Reply:
Should I also move creation of tmpMSI and config before the icon is copied to config directory?

Reply:
That is not required, it has to be done anyway.